<div style="text-align:left;font-size:2em"><span style="font-weight:bolder;font-size:1.25em">SP2273 | Learning Portfolio</span><br><br><span style="font-weight:bold;color:darkred">Loops (Good) Exercises</span></div>

**Exercise 1**

In [9]:
squares_of_odds=[] #where the odd numbers are going to be stored 
for number in range(10): #for numbers 0-9
    if number %2 == 0: #check if the number is even by dividing by 2, no remainders
        continue #if number is even, skip and go to next number
    squares_of_odds.append(number**2) #calculate square and add to list

print(squares_of_odds)


[1, 9, 25, 49, 81]


**Exercise 2**

In [10]:
squears_of_odds=[number**2 for number in range(10) if number % 2 !=0]
print(squares_of_odds)
# % 2 !=0 used to check if the no. is divisible by 2 

[1, 9, 25, 49, 81]


**Exercise 3**

In [6]:
%%timeit

squares_of_odds=[] 
for number in range(10):
    if number %2==0: 
        continue
    squares_of_odds.append(number**2)

453 ns ± 7.06 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)


In [7]:
%%timeit 

squears_of_odds=[number**2 for number in range(10) if number % 2 !=0]

533 ns ± 200 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)


Between the two solutions, the first method was faster.
This is because the code took 453 ns to run with a standard deviation if 7.06 ns per loop 

**Exercise 4**

In [15]:
#initial data 
initial_amount = 100
final_amount = 50
current_amount = initial_amount
years = 0

while current_amount > final_amount:
    current_amount *=0.95 #decreases by 5% each year
    years += 1

print(f'It will take {years} years for the sample to reduce to half its original amount')


It will take 14 years for the sample to reduce to half its original amount


**Exercise 5**

In [20]:
co2_data = [
    (2000, 369.55), (2001, 371.14), (2002, 373.28), 
    (2003, 375.80), (2004, 377.52), (2005, 379.80), 
    (2006, 381.90), (2007, 383.79), (2008, 385.60), 
    (2009, 387.43), (2010, 389.90), (2011, 391.65), 
    (2012, 393.85), (2013, 396.52), (2014, 398.65),
    (2015, 400.83), (2016, 404.24), (2017, 406.55), 
    (2018, 408.52), (2019, 411.44), (2020, 414.24)
]


for i in range(1, len(co2_data)):
    previous_year, previous_level = co2_data[i - 1]
    current_year, current_level = co2_data[i]
    change = current_level - previous_level
    
    if change >= 3:
        print(f"In {current_year}, CO₂ level increased by {change:.2f} ppm compared to {previous_year}.")


In 2016, CO₂ level increased by 3.41 ppm compared to 2015.
